APMrover2/mode.cpp

```diff
@@ -49,56 +64,96 @@ void Mode::set_desired_heading_and_speed(float yaw_angle_cd, float target_speed)
+        g2.motors.set_throttle(constrain_float(throttle_out, g.throttle_min, g.throttle_max));
     }
 
-    if (!reversed && g.braking_percent != 0 && _speed_error < -g.braking_speederr) {
-        // the user has asked to use reverse throttle to brake. Apply
-        // it in proportion to the ground speed error, but only when
-        // our ground speed error is more than BRAKING_SPEEDERR.
-        //
-        // We use a linear gain, with 0 gain at a ground speed error
-        // of braking_speederr, and 100% gain when groundspeed_error
-        // is 2*braking_speederr
-        const float brake_gain = constrain_float(((-_speed_error)-g.braking_speederr)/g.braking_speederr, 0.0f, 1.0f);
-        const int16_t braking_throttle = g.throttle_max * (g.braking_percent * 0.01f) * brake_gain;
-        g2.motors.set_throttle(constrain_int16(-braking_throttle, -g.throttle_max, -g.throttle_min));
-    }
+    // do not attempt to steer
+    g2.motors.set_steering(0.0f);
+
+    // return true once stopped
+    return stopped;
 }
 
+// estimate maximum vehicle speed (in m/s)
+float Mode::calc_speed_max(float cruise_speed, float cruise_throttle)
+{
+    float speed_max;
+
+    // sanity checks
+    if (cruise_throttle > 1.0f || cruise_throttle < 0.05f) {
+        speed_max = cruise_speed;
+    } else {
+        // project vehicle's maximum speed
+        speed_max = (1.0f / cruise_throttle) * cruise_speed;
+    }
+
+    // constrain to 30m/s (108km/h) and return
+    return constrain_float(speed_max, 0.0f, 30.0f);
+}
 
-// calculate pilot input to nudge throttle up or down
-int16_t Mode::calc_throttle_nudge()
+// calculate pilot input to nudge speed up or down
+//  target_speed should be in meters/sec
+//  cruise_speed is vehicle's cruising speed, cruise_throttle is the throttle (from -1 to +1) that achieves the cruising speed
+//  return value is a new speed (in m/s) which up to the projected maximum speed based on the cruise speed and cruise throttle
+float Mode::calc_speed_nudge(float target_speed, float cruise_speed, float cruise_throttle)
 {
-    // get pilot throttle input (-100 to +100)
-    int16_t pilot_throttle = rover.channel_throttle->get_control_in();
-    int16_t throttle_nudge = 0;
-
-    // Check if the throttle value is above 50% and we need to nudge
-    // Make sure its above 50% in the direction we are travelling
-    if ((fabsf(pilot_throttle) > 50.0f) &&
-        (((pilot_throttle < 0) && rover.in_reverse) ||
-         ((pilot_throttle > 0) && !rover.in_reverse))) {
-        throttle_nudge = (rover.g.throttle_max - rover.g.throttle_cruise) * ((fabsf(rover.channel_throttle->norm_input()) - 0.5f) / 0.5f);
+    // return immediately if pilot is not attempting to nudge speed
+    // pilot can nudge up speed if throttle (in range -100 to +100) is above 50% of center in direction of travel
+    const int16_t pilot_throttle = constrain_int16(rover.channel_throttle->get_control_in(), -100, 100);
```

Comment:
Yes, we probably should.  but that's an existing problem so I'll leave that as out-of-scope for this PR.